### PR TITLE
chore: refacto quantized module init + handle no input onnx

### PR DESCRIPTION
**Pull request by jfrery**
closes https://github.com/zama-ai/concrete-ml-internal/issues/4242

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -143,6 +143,8 @@ filterwarnings = [
     "ignore:non-integer arguments to randrange\\(\\) have been deprecated since Python 3\\.10 and will be removed in a subsequent version:DeprecationWarning",
     "ignore:Deprecated call to `pkg_resources.declare_namespace:DeprecationWarning",
     "ignore:The --rsyncdir command line argument and rsyncdirs config variable are deprecated.:DeprecationWarning",
+    "ignore:Converting a tensor to a NumPy array might cause the trace to be incorrect.",
```
Comment (RomanBredehoft) [inline pyproject.toml:146]: I wonder if ignoring these warnings for the whole project is a good thing

File: src/concrete/ml/onnx/ops_impl.py [hunk 2]
```diff
@@ -1495,7 +1496,8 @@ def numpy_cast(data: numpy.ndarray, *, to: int) -> Tuple[numpy.ndarray]:
     if isinstance(data, RawOpOutput):
         return (data.astype(onnx.helper.tensor_dtype_to_np_dtype(to)).view(RawOpOutput),)
 
-    assert_true(to == onnx.TensorProto.BOOL)
+    # Allow both bool and float types
+    assert_true(to in (onnx.TensorProto.BOOL, onnx.TensorProto.FLOAT))
```
Comment (andrei-stoian-zama) [inline src/concrete/ml/onnx/ops_impl.py:1500]: please add an assert message

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]

File: src/concrete/ml/onnx/convert.py [hunk 3]
```diff
@@ -149,6 +150,14 @@ def get_equivalent_numpy_forward_from_torch(
         input_names=arguments,
     )
     equivalent_onnx_model = onnx.load_model(str(output_onnx_file_path))
+
+    # Check if the inputs are present in the model's graph
+    for input_name in arguments:
+        assert_true(
+            any(input_name == node.name for node in equivalent_onnx_model.graph.input),
+            f"Input '{input_name}' is not present in the ONNX model. Please check the onnx graph.",
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/convert.py:158]: I would maybe:
- say `Input '{input_name}' (found in the forward pass)` to be more precise 
- detail a bit more `Please check the onnx graph.` : what should the user do ? if he just builds a torch model, saves it to onnx but then encounters this problem, what could likely be the reason for it ?

File: src/concrete/ml/quantization/quantized_module.py [hunk 4]
```diff
@@ -99,28 +99,39 @@ def __init__(
         # Set base attributes for API consistency. This could be avoided if an abstract base class
         # is created for both Concrete ML models and QuantizedModule
         # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/2899
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module.py:101]: better to move this fixme right above `self.input_quantizers` because it talks about these attributes, no the `self.ordered_module_input_names` and similar

File: src/concrete/ml/quantization/quantized_module.py [hunk 5]
```diff
@@ -99,28 +99,39 @@ def __init__(
...
+
+        all_or_none_params = [
+            ordered_module_input_names,
+            ordered_module_output_names,
+            quant_layers_dict,
+        ]
+        assert_true(
+            all(v is None or v == {} for v in all_or_none_params)
+            or not any(v is None or v == {} for v in all_or_none_params),
+            "All of ordered_module_input_names, ordered_module_output_names, "
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module.py:111]: I would rephrase this by "Please either set all three 'ordered_module_input_names', 'ordered_module_output_names' and 'quant_layers_dict' or none of them" 

also, we should avoid putting `assert` or `assert_true` for user inputs like this and instead add ValueError raises + an associate small test (cf https://github.com/zama-ai/concrete-ml-internal/issues/4232)

File: src/concrete/ml/quantization/quantized_module.py [hunk 6]
```diff
@@ -99,28 +99,39 @@ def __init__(
...
-        assert all([ordered_module_input_names, ordered_module_output_names, quant_layers_dict])
-        self.ordered_module_input_names = tuple(ordered_module_input_names)
-        self.ordered_module_output_names = tuple(ordered_module_output_names)
-
-        assert quant_layers_dict is not None
-        self.quant_layers_dict = copy.deepcopy(quant_layers_dict)
-
-        self.output_quantizers = self._set_output_quantizers()
+        # Initialize output quantizers based on quant_layers_dict
+        if self.quant_layers_dict:
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module.py:131]: I don't understand the modifications in this file, what happens if we don't give any of the inputs like `quant_layers_dict` ? are the `output_quantizers` built later ? I don't see where this is used in the new test 😅

File: tests/torch/test_compile_torch.py [hunk 7]
```diff
@@ -1373,3 +1373,32 @@ def test_mono_parameter_rounding_warning(
...
+            """Forward pass."""
+            zeros = numpy.zeros_like(x)
+            x = x + zeros
+            x = (x > self.threshold).to(torch.float32)
+            x = self.fc1(x)
+            return x
+
+    model = SimplifiedNet()
+
+    with pytest.raises(AssertionError) as excinfo:
```
Comment (RomanBredehoft) [inline tests/torch/test_compile_torch.py:1401]: as said above, this should be a ValueError then !

File: src/concrete/ml/onnx/ops_impl.py (hunk 2 above)
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:1500]: yes not sure where this is used or tested !

File: pyproject.toml (hunk 1 above)
Comment (jfrery) [reply pyproject.toml:146]: It's for pytest only

File: src/concrete/ml/quantization/quantized_module.py (hunk 6 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:131]: This is just some refacto so that we actually have default values for all attributes. It fixes nothing though. The test would be to check that attribute at indeed initialized to something but I find this pretty irrelevant.

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py:1401]: Oh yeah I forgot about this. Will fix.

File: src/concrete/ml/onnx/ops_impl.py (hunk 2 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:1500]: Done

File: src/concrete/ml/onnx/convert.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/onnx/convert.py:158]: > say Input '{input_name}' (found in the forward pass) to be more precise

```
f"'{input_name}' is missing in the ONNX graph after export. Verify the forward pass for issues."
```

But  about

> detail a bit more Please check the onnx graph. : what should the user do ? if he just builds a torch model, saves it to onnx but then encounters this problem, what could likely be the reason for it ?

I have no idea. For the test we added it was to not use numpy in the forward but I am sure there are tons of different ways to make an onnx input disappear and so as much reasons.

File: src/concrete/ml/quantization/quantized_module.py [hunk 8]
```diff
@@ -99,28 +99,39 @@ def __init__(
...
+        )
+        self.ordered_module_output_names = (
+            tuple(ordered_module_output_names) if ordered_module_output_names else ()
+        )
+        self.quant_layers_dict = (
+            copy.deepcopy(quant_layers_dict) if quant_layers_dict is not None else {}
+        )
+        self.input_quantizers: List[UniformQuantizer] = []
+        self.output_quantizers: List[UniformQuantizer] = []
+        self.fhe_circuit: Union[None, Circuit] = None
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:125]: Yes sure

File: src/concrete/ml/quantization/quantized_module.py (hunk 5 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:111]: Sure I rephrased.

About the second part, it's not really user inputs here. Users use the QuantizedModule through the PTQ. I can add a test anyway yes.

**Review by andrei-stoian-zama — APPROVED** [review]
perfect!

File: pyproject.toml (hunk 1 above)
Comment (RomanBredehoft) [reply pyproject.toml:146]: ah yes indeed, but still these warnings seem pretty important no ?

File: src/concrete/ml/onnx/ops_impl.py (hunk 2 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py]: but this test itself is meant for another feature. Could you maybe add a specific test that only does this `to(float)` ?

Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py]: I have a test that check different cast yes.

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py]: Should be good.

File: src/concrete/ml/onnx/ops_impl.py (hunk 2 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py]: ah yes sorry didn't see, great then thanks !

File: src/concrete/ml/quantization/quantized_module.py (hunk 6 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:132]: The only different is that now, we initialize all attributes. Before we were just returning so some attributes were not present. It was just bad practice. This refacto should not change anything. Testing that attributes are properly initialized could be done but then why not doing it everywhere.

File: src/concrete/ml/onnx/convert.py (hunk 3 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/convert.py:158]: ah yes I see, well ok we can keep it that way I guess. anyway yes for the first point I like your suggestion

File: src/concrete/ml/onnx/ops_impl.py [hunk 9]
```diff
@@ -1495,7 +1495,12 @@ def numpy_cast(data: numpy.ndarray, *, to: int) -> Tuple[numpy.ndarray]:
...
-    assert_true(to == onnx.TensorProto.BOOL)
+    allowed_types = (onnx.TensorProto.BOOL, onnx.TensorProto.FLOAT, onnx.TensorProto.DOUBLE)
+    assert to in allowed_types, (
+        f"Invalid 'to' data type: {onnx.TensorProto.DataType.Name(to)}. "
+        f"Only {', '.join(onnx.TensorProto.DataType.Name(t) for t in allowed_types)}"
+        "are allowed for casting."
+    )
 
     # Will be used for traced values
     return (data.astype(numpy.float64),)
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:1506]: not sure what is happening here then : we check that `to` is a bool, float or double, but we only cast to double without talking `to` into account ?

File: pyproject.toml (hunk 1 above)
Comment (jfrery) [reply pyproject.toml:146]: Yes! IMO these warning should be runtime errors. Not sure why torch allows exporting to onnx with broken forward.

But in pytest, warnings == failures. So we have to filter these.

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
a few remaining comments from my side !

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (RomanBredehoft) [reply tests/torch/test_compile_torch.py]: actually you were right not to change anything here this was a legit assert my bad

File: src/concrete/ml/quantization/quantized_module.py (hunk 5 above)
Comment (RomanBredehoft) [reply src/concrete/ml/quantization/quantized_module.py]: it's still a public class that we even can serialize like any built-in models. But actually I meant "inputs" more than 'user inputs", asserts should not really be used to check input value formats

File: src/concrete/ml/onnx/ops_impl.py (hunk 9 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:1506]: Yes it's pretty weird indeed. And before it was a cast to bool and then we casted to float64 so even weirder.

I think the idea is that we want to be able to support cast (i.e. not break because of a cast in the torch forward) but we don't want to change the FHE semantic. Casting to anything else than float64 is dangerous as it could be interpreted by concrete-python improperly and close a LUT.

Actually as @andrei-stoian-zama stated casting to float64 could be potentially dangerous as well if inputs where integers as it would mean opening a LUT (while the cast actually as no real effect in FHE).

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py]: not sure I understood? I actually think you were right? Why not using the feature from pytest.raises instead of creating a new line with an assert?

File: src/concrete/ml/onnx/ops_impl.py (hunk 9 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py:1506]: ah yes actually now that I think of it, I remember already having some issues with that `numpy_cast` in the past

it feels like there's something we should do about your last point though !

File: src/concrete/ml/onnx/convert.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/onnx/convert.py]: I changed the message. Should be good.

File: src/concrete/ml/onnx/ops_impl.py (hunk 9 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:1506]: But would this kind of check integer be for the numpy cast only? I think this is a more general concern. Also we have function that do work on integers (tree specific numpy method) and some function that work on both (numpy_gemm).

File: pyproject.toml (hunk 1 above)
Comment (RomanBredehoft) [reply pyproject.toml:146]: but if these warnings should be considered warnings, then maybe it's best if we add a `with pytest.raise` for them instead of ignoring them no ? 🤔

File: src/concrete/ml/onnx/ops_impl.py (hunk 9 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py:1506]: yes yes for sure this is more of a general concern, maybe we should summarize this in a dedicated issue ?

Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:1506]: Will do

File: pyproject.toml (hunk 1 above)
Comment (fd0r) [reply pyproject.toml:146]: This isn't a warnings raised from CML so I see no issue in just ignoring it.
Using `with pytest.raise` would mean that we want the warning to be raised.

File: src/concrete/ml/onnx/ops_impl.py (hunk 9 above)
Comment (fd0r) [reply src/concrete/ml/onnx/ops_impl.py:1506]: We could maybe have something that does:
```
if not int: return float
if int: return int
```
or smth like that

**Review by fd0r — APPROVED** [review]
Fine with me as is

File: pyproject.toml (hunk 1 above)
Comment (RomanBredehoft) [reply pyproject.toml:146]: yes ok you're right, fine  for me then !

**Review by RomanBredehoft — APPROVED** [review]
fine with we thanks a lot ! just dpn't forget to create an issue for the `numpy_cast` part !
